examples/tutorials/mnist_pytorch/hotpatches/harness_hotpatch.py

```diff
@@ -0,0 +1,688 @@
+            if last_measurement_time is None:
+                last_measurement_time = time.time()
+                batch_start_time = time.time()
+                network_throughput_collector.reset()
+                disk_collector.reset()
+
+
+
+
+            # Check if it is time to take a new measurement
+            if time.time() - last_measurement_time > self.MEASUREMENT_INTERVAL:
+                self.log("Taking new measurement")
+                immutable_batch_idx = self.current_batch
+                self.log("Taking new measurement - cpu")
+                cpu_util_measurement = cpu_util_collector.measure(immutable_batch_idx)
+                self.log("Taking new measurement - gpu")
+                gpu_util_measurement = gpu_util_collector.measure(immutable_batch_idx)
+                self.log("Taking new measurement - network")
+                net_thru_sent_measurement, net_thru_recv_measurement = network_throughput_collector.measure(immutable_batch_idx)
+                self.log("Taking new measurement - memory")
+                free_memory_measurement = free_memory_collector.measure(immutable_batch_idx)
+                self.log("Taking new measurement - disk")
+                disk_read_thru_measurement, disk_write_thru_measurement, iops_measurement = disk_collector.measure(immutable_batch_idx)
+
+                self.current_metric_batch[GPU_UTIL_METRIC].extend(gpu_util_measurement)
+                self.current_metric_batch[NET_THRU_SENT_METRIC].append(net_thru_sent_measurement)
+                self.current_metric_batch[NET_THRU_RECV_METRIC].append(network_throughput_collector)
+                self.current_metric_batch[DISK_IOPS_METRIC].append(iops_measurement)
+                self.current_metric_batch[DISK_THRU_READ_METRIC].append(disk_read_thru_measurement)
+                self.current_metric_batch[DISK_THRU_WRITE_METRIC].append(disk_write_thru_measurement)
+                self.current_metric_batch[FREE_MEM_METRIC].append(free_memory_measurement)
+                self.current_metric_batch[SIMPLE_CPU_UTIL_METRIC].append(cpu_util_measurement)
+                self.log("Finished taking measurement")
+
+
+            # Check if it is time to flush the batch and start a new batch
+            if time.time() - batch_start_time > self.FLUSH_INTERVAL:
+                self.log("Completed a batch")
+                self.enqueue_for_async_send(self.current_metric_batch)
+                self.current_metric_batch = {
+                    GPU_UTIL_METRIC: [],
+                    NET_THRU_SENT_METRIC: [],
+                    NET_THRU_RECV_METRIC: [],
+                    DISK_IOPS_METRIC: [],
+                    DISK_THRU_READ_METRIC: [],
+                    DISK_THRU_WRITE_METRIC: [],
+                    FREE_MEM_METRIC: [],
+                    SIMPLE_CPU_UTIL_METRIC: []
+                }
+                batch_start_time = time.time()
+
+
+
+    def update_current_batch(self, new_current_batch):
```

Comment:
where/how is this going to be called?

Reply:
It will be called by the controller. The code has been updated to do that for the PyTorch controller